### DISCUSSION TITLE
Is this a false negative about Self

### DISCUSSION THREAD

**Discussion by lexdene** [python/typing#1389] [Q&A] [answered]
code (src/main.py):

    from typing import Self

    class Foo:
        key: str
        name: str

        def __init__(self, key: str, name: str):
            self.key = key
            self.name = name

        @classmethod
        def get(cls, key: str) -> Self:
            return _data[key]

    _data: dict[str, Foo] = {
        '0': Foo('0', 'hello'),
        '1': Foo('1', 'world'),
    }

script:

    mypy -V
    mypy src

expect output:

    mypy 1.2.0 (compiled: yes)
    Success: no issues found in 1 source file

actual output:

    mypy 1.2.0 (compiled: yes)
    src/main.py:14: error: Incompatible return value type (got "Foo", expected "Self")  [return-value]
    Found 1 error in 1 file (checked 1 source file)

additional infos:

    $ pip show mypy
    Name: mypy
    Version: 1.2.0
    Summary: Optional static typing for Python
    Home-page: http://www.mypy-lang.org/
    Author: Jukka Lehtosalo
    Author-email: jukka.lehtosalo@iki.fi
    License: MIT License
    Location: ~/venv/lib/python3.11/site-packages
    Requires: mypy-extensions, typing-extensions
    Required-by: 
    $ python --version
    Python 3.11.1
    $ lsb_release -a
    No LSB modules are available.
    Distributor ID:	Ubuntu
    Description:	Ubuntu 20.04.6 LTS
    Release:	20.04
    Codename:	focal

**Comment by JelleZijlstra** [comment]
The error is correct. If you have a subclass of `Foo`, this function will still return `Foo` objects, not instances of the subclass.

**Reply by lexdene** [reply]
How should I tell mypy that `_data[key]` is an instance of `Foo` and will never be an instance of a subclass of `Foo`?

**Reply by JelleZijlstra** [reply, marked as answer]
Your method should be marked as returning `Foo`.

**Reply by deliro** [reply]
It's worth noting that marking a class as `@final` is also a solution (IMO a better one)
